osd/rados/rgw/cephfs: Modernize cls interface with compile time safety

aainscow opened the pull request:
This is a preparation PR for EC direct reads, but is also provides protection for balanced reads. 

This PR does the following:

1. Adds a new C++ interface, with compile time safety, which prevents any code from attempting to call an exec which may do a write without marking the OP as write. 
2. Deletes any C++ interface which can be used unsafely. 
3. Enforces that IOCtx::exec() and IOCtx::aio_exec() are read only. 
4. Refactors all clients within Ceph to use the new interface. 
5. Deletes unused clients (cls_acl, cls_crypto and key_value_store) + associated tests


Following reviews of earlier versions and discussion at the CDM, we are NOT:

- Deprecating the C interfaces and associated bindings to other languages. 
- Providing any safety in the client for the C / other languages interfaces

A **later** PR will add policing to the OSD:
- Umbrella clients sending an illegal exec (i.e. one where a read contains a write) will be failed by the OSD. 
- Pre-Umbrella clients sending an illegal exec will cause a health warning to be generated, encouraging the user to update their client. 

Tracker: https://tracker.ceph.com/issues/73986


<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)

You must only issue one Jenkins command per-comment. Jenkins does not understand
comments with more than one command.
</details>

bill-scales reviewed (commented):
How are you going to identify the set of class methods that need to be annotated as read only to get the performance benefit of direct reads?

Testing different types of I/O and looking for ops that don't get treated as direct reads is one method, but unless you have very good test coverage you might miss something, especially if an exec call is only added to ops when a specific feature (e.g. rbd mirroring) is in use.

I suspect it might be easier to just modify as many of the cls client code functions (and implementations in neorados) to add the read only flag rather than try and find cases via testing. Maybe that should be a 2nd PR?

aainscow commented:
jenkins test api

cbodley commented:
is there some way the rados client library can base this direct-read decision solely on the use of `ObjectWriteOperation` vs `ObjectReadOperation`, without requiring applications to mark their calls to exec?

that wouldn't help with `IoCtx::exec()`, but https://tracker.ceph.com/issues/65889 led to two commits https://github.com/ceph/ceph/pull/56180/commits/af176311c4768fad3d6ca630e4f769e3012d2283 and https://github.com/ceph/ceph/pull/61471/commits/4612195b9c3b9a86142cb2a4b04366ac14840295 that replaced some `IoCtx::exec()` with calls to `ObjectWriteOperation::exec()`. if this implies `IoCtx::exec()` should be used only for reads, then that could be flagged as read-only and enforced on the osd with EIO

aainscow commented:
> is there some way the rados client library can base this direct-read decision solely on the use of `ObjectWriteOperation` vs `ObjectReadOperation`, without requiring applications to mark their calls to exec?
> 
> that wouldn't help with `IoCtx::exec()`, but https://tracker.ceph.com/issues/65889 led to two commits [af17631](https://github.com/ceph/ceph/commit/af176311c4768fad3d6ca630e4f769e3012d2283) and [4612195](https://github.com/ceph/ceph/commit/4612195b9c3b9a86142cb2a4b04366ac14840295) that replaced some `IoCtx::exec()` with calls to `ObjectWriteOperation::exec()`. if this implies `IoCtx::exec()` should be used only for reads, then that could be flagged as read-only and enforced on the osd with EIO

There is currently no policing in librados, or the OSD that the client correctly used ObjectReadOperation vs ObjectWriteOperation.  I did consider simply overriding the ObjectOperation::exec() functions in ObjectReadOperation and assuming they are all read only.  However, if any client has mistakenly used an ObjectReadOperation::exec() incorrectly, then this change would prevent that client from working. I considered this to be too high risk. 

I am open to arguments to the contrary on this, as overrides would be a really neat way of implementing this. 

Unfortunately, this would not work for NEORADOS, where I can't see a way of being explicit about read vs write.

cbodley commented:
thanks @aainscow,

> I did consider simply overriding the ObjectOperation::exec() functions in ObjectReadOperation and assuming they are all read only.

i was hoping we could handle this in the `ObjectReadOperation` overloads for `librados::IoCtx::operate()`/`aio_operate()` (and `neorados::RADOS::execute()` for `ReadOp`) without needing to treat `exec()` as a special case

> However, if any client has mistakenly used an ObjectReadOperation::exec() incorrectly, then this change would prevent that client from working. I considered this to be too high risk.

i think it would be nice to catch any misuse*. ceph components may not have enough test coverage to rely on osd enforcement to catch everything, but a code audit probably could. worth discussing, at least?

edit: *especially now that https://github.com/ceph/ceph/pull/56180 relies on this for local/balanced read affinity

cbodley commented on src/include/rados/librados.hpp:
```diff
@@ -343,6 +345,20 @@ inline namespace v14_2_0 {
+        exec_ro(method.cls, method.name, std::forward<Args>(args)...);
+      } else {
+        exec(method.cls, method.name, std::forward<Args>(args)...);
+      }
```
as-is, we wouldn't catch misuses like this:
```c++
void inc_version(ObjectOperation& op)
{
  bufferlist in;
  op.exec(::cls::version::MethodInc, in);
}

void foo()
{
  ObjectReadOperation op;
  inc_version(op);
```
probably better for `ObjectOperation::exec()` to only support reads, with a new `ObjectWriteOperation::exec()` that supports both reads and writes

aainscow committed:
librados: enforce compile-time read/write semantics for class methods

Status Quo:
Librados currently uses string-based identifiers for RADOS class method calls
(exec) in both ObjectReadOperation and ObjectWriteOperation. The API does
not distinguish between methods that modify object state and those that
are read-only.

Problem:
This lack of semantic enforcement has become a critical safety gap with the
introduction of EC Direct Reads. Because the EC direct read path is optimized
for performance, it may process the same read-only operation multiple times
(e.g., across different shards or during retries).

Previously, while technically a violation of API semantics, a "read-that-writes"
would often function correctly because operations were processed in-order and
without retries on the read path. However, in the context of EC direct reads,
executing a non-idempotent write through the read path poses a silent and
critical risk to data integrity.

Solution:
Introduce a template-based trait system to encode method semantics into the
API. By tagging methods with traits (MethodRead or MethodWrite), the compiler
now prevents state-modifying methods from being added to ObjectReadOperation.
This ensures that only side-effect-free methods reach the EC direct read
path, eliminating the risk of accidental double-writes at the source.

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

aainscow committed:
cls: refactor class directory and tests to use trait-based API

Migrate existing internal RADOS classes and the associated test infrastructure
to utilize the new template-based exec() interface. While the previous
commit introduced the mechanism, this change applies it across the
codebase to eliminate legacy string-based calls in internal logic.

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

aainscow committed:
cls: Remove legacy ACL/Crypto and key_value_store class methods.

The osd code was not present, the radosacl tool did nothing useful and the scratchtoolpp
was testing something that doesn't exist.

key_value_store is optionally compiled into the key store, but seems to serve no useful purpose.

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

aainscow committed:
test: Fix up librados/neorados tests

There much mocking of exec going on in these tests. An extensive
search and replace was required to find them all.

Note that the failure looks like memory corruption when manipulating
the ops array - but the actually issue is the mocking not intercepting
the exec_impl call (and similar) correctly.

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

aainscow committed:
rgw: migrate to using non-deprecated class call functions

Update the Rados Gateway driver to utilize the new trait-based class
method interface. This replaces legacy string-based exec() calls with
the typesafe template-based alternatives.

Key Changes:

Updated reset_stats in buckets.cc to use cls::user::method::reset_user_stats2.

Updated RGWRadosBILogTrimCR in rgw_cr_rados.cc to use cls::rgw::method::bi_log_trim.

This change ensures that RGW operations are correctly categorized as
reads or writes at compile-time, aligning with the broader effort to
improve RADOS class method safety and OSD routing efficiency.

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

aainscow commented:
This is a major refactor, so is accompanied by a re-write and rename of the PR. 

This is quite a big change to the API. Backward compatibility is maintained.

cbodley commented on src/include/rados/librados.hpp:
```diff
@@ -14,6 +14,8 @@
 #include "librados_fwd.hpp"
 #include "rados_types.hpp"
 
+#include "cls/cls_traits.h"
```
cls_traits.h and cls_flags.h should probably be under src/include/rados and installed with the other headers by src/include/CMakeLists.txt for inclusion in packages like librados-devel and libradospp-devel

cbodley commented on src/include/rados/objclass.h:
```diff
@@ -85,8 +82,26 @@ typedef int (*cls_method_cxx_call_t)(cls_method_context_t ctx,
  * @param class_call
  * @param handle
  */
-extern int cls_register_cxx_method(cls_handle_t hclass, const char *method, int flags,
```
we package objclass.h in rados-objclass-devel so there's some expectation of api stability here

aainscow replied:
There is also the python interface to consider. What should we do with that?

aainscow commented on src/include/rados/librados.h:
```diff
@@ -3066,6 +3066,10 @@ CEPH_RADOS_API void rados_write_op_zero(rados_write_op_t write_op,
  *
+ * Use this for reads as well as writes. The rados_read_op_exec has been
+ * deprecated as there is no client-side checking for the chosen cls method
+ * actually being a read vs write.
```
I would argue that the "Write" version is the replacement.  A C client will have to perform writes in order to perform any execs, or use the one-shot exec() call (which does a write)

aainscow replied:
I think this is what you suggest in the next field.  I totally agree that cls API is kind of a "middle" API - not really intended for clients to consumer directly.  As mentioned below, the python API also exists (this calls ioctx->exec or ioctx->aio_exec.

rzarzynski commented on src/include/rados/cls_flags.h:
```diff
@@ -0,0 +1,5 @@
+
+#define CLS_METHOD_RD       0x1 /// method executes read operations
+#define CLS_METHOD_WR       0x2 /// method executes write operations
+#define CLS_METHOD_PROMOTE  0x8 /// method cannot be proxied to base tier
```
Huh, sounds like it's because the cache tiering.

rzarzynski commented on src/objclass/class_api.cc:
```diff
@@ -66,6 +66,15 @@ int cls_register_cxx_method(cls_handle_t hclass, const char *method,
 }
 
+// deprecated
+int cls_register_cxx_method(cls_handle_t hclass, const char *method,
```
We'll want eventually to switch all our _plugins_ into the new interface. If external ones can be neglected (I would argue for that), the entire deprecated registration machinery can be dropped.

rzarzynski commented on src/include/neorados/RADOS.hpp:
```diff
@@ -352,22 +355,64 @@ class Op {
   void assert_exists();
   void cmp_omap(const std::vector<cmp_assertion>& assertions);
 
+  [[deprecated("in favor of compile-time safe ops")]]
```
Yeah, in contrast to _objclass_ are undoubtedly public.

rzarzynski commented on src/cls/version/cls_version_types.h:
```diff
@@ -12,6 +12,7 @@
 #include "common/Formatter.h"
 #include "include/encoding.h"
 #include "include/types.h"
+#include "include/rados/cls_traits.h"
```
Do  we really need the include for this particular file?

rzarzynski commented on src/test/librados/misc_cxx.cc:
```diff
@@ -924,3 +926,37 @@ TEST_F(LibRadosMiscPP, Conf) {
+//   cls_version_inc_op call;
+//   encode(call, bl2);
+//   read1.exec_ro("version", "inc", bl2);
+//   EXPECT_EQ(-EIO, ioctx.operate("foo", &read1, nullptr));
```
I wonder whether to `EIO` is the best error code for a misbehaving client. `EINVAL`?

rzarzynski commented on src/test/librados/misc_cxx.cc:
```diff
@@ -924,3 +926,37 @@ TEST_F(LibRadosMiscPP, Conf) {
+//   ObjectReadOperation read1;
+//   cls_version_inc_op call;
+//   encode(call, bl2);
+//   read1.exec_ro("version", "inc", bl2);
```
Exec-read-only but without the tag, so no compile-time safety. ACK.

rzarzynski commented on src/test/librados/misc_cxx.cc:
```diff
@@ -924,3 +926,37 @@ TEST_F(LibRadosMiscPP, Conf) {
+  decode(all_features, iter);
+  // make sure *some* features are specified; don't care which ones
+  ASSERT_NE(all_features, (unsigned)0);
+}
```
Would be nice to prove _exec_ro with tag_ fails at compile-time. Should be possible with GTest.

rzarzynski commented on src/include/neorados/RADOS.hpp:
```diff
@@ -407,6 +407,24 @@ class Op {
+   *
+   * This Op may be a read, so only read execs are permitted with this interface
+   *
+   * @param method the method as defined in cls/<class>/cls_<class>_ops.h
```
This basically says that a _Ceph Class_ should exist within the Ceph repo. I'm not going to disagree ;-).

rzarzynski commented on src/crimson/osd/osd_operations/client_request.cc:
```diff
@@ -193,6 +193,20 @@ ClientRequest::interruptible_future<> ClientRequest::with_pg_process_interruptib
+      co_return;
+    }
+    // Legacy clients may rely on this behaviour, so do not drop or fail.
+  }
```
CC: @Matan-B.

rzarzynski reviewed (commented):
Generally LGTM!

rzarzynski commented:
@aainscow: just as side note: if there is no ticket in the _RADOS_ tracker, would you mind creating one?

gregsfortytwo commented:
This PR has had some iterations it looks like, but this comment concerns me @aainscow:
> Refactors "IOCTX" interface to flag all execs as writes. This is safe, since reads are always permitted as part of writes.

Does this mean it’s no longer possible to issue class operations as reads at all? Writes are a lot more expensive than reads (they have to be ordered!) so this isn’t great.

I also not sure about @rzarzynski’s thoughts about out of tree users — I thought we had some. But the one in my head is @dalgaaf ’s dovecot email plugin and it looks like that is just using librados, so maybe it’s fine.

gregsfortytwo commented:
Pinging @athanatos and @batrick for attention on this big change as well.

batrick reviewed (changes requested):
The commit messages must have exposition on (a) the status quo; (b) the problem to be solved; (c) how you're solving it. Some code examples in the commit message would also be helpful.

There should be a PendingReleaseNote! Documentation updates are also necessary.

I don't have time to go through all of the various comments on this but I was also alarmed by Greg's comment on Slack that librados exec methods methods can not be read-only?

> I would argue that the "Write" version is the replacement. A C client will have to perform writes in order to perform any execs, or use the one-shot exec() call (which does a write)

A "write" version would require appropriate authorizations, no?

Is the whole point of this to make sure the client directs the op to the primary OSD?

aainscow commented:
Latest update does a better job of removing the radosacl executable, which was causing the windows build to fail.   

NOte that the cls_acl was compiled, but nothing ever called it, so the radosacl exec would have failed.

adamemerson reviewed (changes requested):
Just get the `using namespace` out of `version.h`.

aainscow commented on src/include/rados/cls_traits.h:
```diff
@@ -0,0 +1,38 @@
+
+template <typename Tag, typename ClassDef>
+struct ClsMethod {
+  const char* cls;
```
Yes, with a bit of code churn. I have submitted a commit - if you approve, I will squash it into older commits.

adamemerson reviewed (approved):
Looks good to me! Drop the `string_view` change. Everything else is fine and should be good compile-only.

aainscow commented on src/include/neorados/RADOS.hpp:
```diff
@@ -408,7 +408,7 @@ class Op {
     static_assert(FlagTraits<Tag>::is_readonly,
       "Attempt to call a non-readonly class method as part of a non-write op ");
-    exec_impl(method.cls, method.name, inbl, std::forward<Args>(args)...);
+    exec_impl(method.cls.data(), method.name.data(), inbl, std::forward<Args>(args)...);
```
Will do. Thanks for the help.

rzarzynski commented on src/include/rados/cls_traits.h:
```diff
@@ -0,0 +1,49 @@
+};
+
+template <typename Tag>
+constexpr bool is_safe_for_ro_v = FlagTraits<Tag>::is_readonly;
```
Not a pure-C header.

aainscow committed:
*: Update PendingReleaseNotes for librados.hpp CLS breaking API change

Signed-off-by: Alex Ainscow <aainscow@uk.ibm.com>

batrick reviewed (approved):
Great work!

JonBailey1993 commented:
Rados approved: https://tracker.ceph.com/projects/rados/wiki/MAIN#httpstrackercephcomissues75385
